CORE-1547 Sortable tree headers

Swizec opened the pull request:
Users can click on tree headers to sort by that column.

Some fields were too convoluted to implement, so I disabled them. An example of this is sorting by authorizations in people trees.

hyperNURb commented on src/ggrc/assets/javascripts/controllers/tree_view_controller.js:
```diff
@@ -443,6 +445,12 @@ CMS.Controllers.TreeLoader("CMS.Controllers.TreeView", {
+
+              can.bind.call(that.element.parent().find('.widget-col-title'),
+                            'click',
+                            function (event) {
```
`Event` is a keyword, could you please use something different, since this can be error prone?

Swizec replied:
Event is not in fact a keyword. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar

It is not even a future reserved word.

hyperNURb commented on src/ggrc/assets/javascripts/controllers/tree_view_controller.js:
```diff
@@ -443,6 +445,12 @@ CMS.Controllers.TreeLoader("CMS.Controllers.TreeView", {
+                            'click',
+                            function (event) {
+                              that.sort(event);
+                            });
```
Also you should use `.bind(this)` here. or even better `can.bind.call(that.element.parent().find('.widget-col-title'), 'click', this.sort.bind(this)`

Swizec replied:
I'm just following the style from the rest of this function. But yes, the whole function should be updated.

hyperNURb commented on src/ggrc/assets/javascripts/controllers/tree_view_controller.js:
```diff
@@ -873,6 +881,41 @@ CMS.Controllers.TreeLoader("CMS.Controllers.TreeView", {
+
+  , sort: function (event) {
+      var $el = $(event.currentTarget),
+          key = $(event.currentTarget).data("field");
```
You already have stashed `$el` use it instead of making a new _jQuery_ object.

Swizec replied:
No I don't have stashed $el. That's a different element.

hyperNURb replied:
```
var $el = $(event.currentTarget),
      key = $(event.currentTarget).data("field");
``
 to

```
var $el = $(event.currentTarget),
      key = $el.data("field");
``

Swizec replied:
Oh, didn't even notice that. My bad :)

hyperNURb commented on src/ggrc/assets/javascripts/controllers/tree_view_controller.js:
```diff
@@ -873,6 +881,41 @@ CMS.Controllers.TreeLoader("CMS.Controllers.TreeView", {
       this.display_prefs.save();
     }
+
+  , sort: function (event) {
```
As before, `Event` is a keyword and this is possibly error prone.

Swizec replied:
As before, event is not a keyword. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar

hyperNURb commented on src/ggrc/assets/javascripts/models/cacheable.js:
```diff
@@ -1077,6 +1077,40 @@ can.Model("can.Model.Cacheable", {
+  // owners.0.name|email -> 
+  //  firstnonempty this.owners[0].reify().name this.owners[0].reify().email
+  get_deep_property: function get_deep_property (descriptor, val) {
+    val = typeof val === "undefined" ? this : val;
```
Is there a reason you don't use `val = val || this;`?

Swizec replied:
@Smotko doesn't like it.

hyperNURb replied:
Nope, he doesn't have problem with that, he has problem with `val || (val = this);` but please use `val = val || this`.
@zidarsk8 likes to use the concept you used now, but he is python dev :tongue:

Swizec replied:
Actually, I do have to use this because I could potentially get valid falsy values in there and I am specifically checking just for undefined.

ie. this is only valid on the first call of `get_deep_property`, not when it recurses.

hyperNURb replied:
Ok :+1: It's a valid reason

hyperNURb commented on src/ggrc/assets/javascripts/models/cacheable.js:
```diff
@@ -1077,6 +1077,40 @@ can.Model("can.Model.Cacheable", {
+    
+    if (!descriptor || !descriptor.length) {
+      return undefined;
+    }else if (val[descriptor]) {
```
Code style `} else if …`

hyperNURb commented on src/ggrc/assets/javascripts/models/cacheable.js:
```diff
@@ -1077,6 +1077,40 @@ can.Model("can.Model.Cacheable", {
+      return undefined;
+    }else if (val[descriptor]) {
+      return val[descriptor];
+    }else{
```
`} else {`

hyperNURb commented on src/ggrc/assets/javascripts/models/cacheable.js:
```diff
@@ -1077,6 +1077,40 @@ can.Model("can.Model.Cacheable", {
+
+        if (typeof val[key] === "undefined" || val[key] === null) {
+          return undefined;
+        }else{
```
`} else {`

Swizec commented:
Fixed styling issues. It also turns out @zidarsk8 and I implemented almost the same function to access deep properties, but I'm not sure if they're exactly the same. His seems very reliant on the idea that it's part of filters.

Smotko commented:
@Swizec todo: 

- [ ] Sync up with @zidarsk8 if we need two functions with similar functionality

@vladan-mitevski / @arraystudio todo:

- [ ] We should use a pointer icon on tree headers (clicking with the cursor feels weird)
- [ ] There is no indication on which column is being sorted (I think bolding out the active column header would be good, but you might have a better idea).
- [ ] There is no indication whether the sorting in DESC or ASC (flipping the icon?)

vladan-mitevski commented:
@Smotko  There should be carot rotation based on ASC or DESC sort. For second task we need to see with @arraystudio what to do cause the requirements said there is no sorting indication on page load, so only after user clicks on that column title arrow will be shown for one type and if you click it again it will flip the icon.

Swizec commented:
@smotko please run build_compass, the indicators are there.

On 20 April 2015 at 03:42, Vladan Mitevski <notifications@github.com> wrote:

> @Smotko <https://github.com/Smotko> There should be carot rotation based
> on ASC or DESC sort. For second task we need to see with @arraystudio
> <https://github.com/arraystudio> what to do cause the requirements said
> there is no sorting indication on page load, so only after user clicks on
> that column title arrow will be shown for one type and if you click it
> again it will flip the icon.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/ggrc-core/pull/2616#issuecomment-94419251>.
>

Swizec commented:
@Smotko confirmed, everything works, you just forgot to compile CSS :)
And we've agreed with @zidarsk8 that the functions are different.

@vladan-m can you please resolve the merge conflict in icons.png?

vladan-m commented:
@Swizec There are more conflicts except icons.png. Should I fix just icons.png?

Swizec commented:
@vladan-m are there? Sorry, I only noticed icons.png. Can you fix whatever you can and I'll handle any JS conflicts.

vladan-m commented:
@Swizec OK. All other conflcits are in `tree_header.mustache` files. I'll fix what I can. :smiley:

vladan-m commented:
@Swizec I've fixed merge conflicts except for JS. In this file `javascripts/controllers/tree_view_controller.js` on line 512 you will see start of the conflict. Can you check it out and remove wrong code?

Swizec commented:
Will do, tomorrow. It's getting late here.

Thanks for the help.

On 23 April 2015 at 00:53, Vladan Mitevski <notifications@github.com> wrote:

> @Swizec <https://github.com/Swizec> I've fixed merge conflicts except for
> JS. In this file javascripts/controllers/tree_view_controller.js on line
> 512 you will see start of the conflict. Can you check it out and remove
> wrong code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/ggrc-core/pull/2616#issuecomment-95480062>.
>

vladan-m commented:
@Swizec  Sure. No problemo :smile:

Smotko commented:
@Swizec unfortunately you are going to have to move this PR against develop. epic/mango branch is no more.
